README
= Rhotoalbum

version 0.9

January 17, 2011

Ondrej Jaura  mailto:ondrej@valibuk.net
Viktor Zigo  mailto:viz@alephzarro.com (contributor)


== I. Description

Rhotoalbum is a photo album generator programmed in Ruby.

It works with a directory structure, it produces static web content and it does not require any database.

Rhotoalbum generates thumbnails and the index.html (rss.xml) file in each subdirectory. There is a possibility to add titles and descriptions of the images.

== II. Web site

Please see the Rhotoalbum web site http://rhotoalbum.valibuk.net for futher information such as installation, configuration, best practices or F.A.Q.



changelog.txt
* 0.9 Do not open

20110117 - Ondrej Jaura

- bugfix #25937: Too many open files. (thnx to Mooffie)

* 0.8 We are back

20100719 - Ondrej Jaura

- album cover image support - cover.jpg
- reverse generating to provide highlight.jpg also for parent albums
- cleanhighlight command added
- (web page reworked)

20100704 - Ondrej Jaura

- effects for albums added
- polaroid_stack effect added
- web page updated (album effects, rmagick dependency)

20100703 - Ondrej Jaura

- fix duplicate image entries, bug #28352

* 0.7 Shouldn't this be 1.0? (A modest release label :)
[not officially released]

20081025 - Viktor Zigo

- text encoding (HTML entities)
- text in slideshow
- optional effects (optionally requires RMagick)

20080916 - Ondrej Jaura

- bugfix #21362: Generate the date_time from the exifr data instead of filename. (thnx to Scott Likens)
- bugfix #21363: If there is no focal length in the image (specified in exif) then don't display any focal length at all. (thnx to Scott Likens)
- only existing EXIF properties are added to the photo description

20080829 - Viktor Zigo

- options.yml can be defined for each album (inheriting from parent) = each album can have its own properties
- optional 'require rubygems' (backward compatibility)
- new command 'rebuild' = executes sequence 'cleanindex', 'text', 'generate'; suitable for editing of existing album
- the 'text' command will merge all the new images to the already existing descrtiption.txt
- media RSS 2.0 photo feed support (best viewed by http://www.cooliris.com/); RSS per each album; N most recent photos
- RSS icon in the menu
- support for pagination, option: maxPerPage  -1 means no pagination; any other number enables album pagination
- option: descriptionAsName [false] - the text descriptions are used as image/album names instead of the plain file/directory name
- option: debug [false] - makes the generation more verbose
- option: nonrecursive execution [false]
- option: google_analytics [off] - generate google analytics code block option
- polished console output (more readable and less verbose)
- some code cleaning

* 0.6  More, more, more

20080610 - Ondrej Jaura

- wrong parameters order for the ln -s example in the README file (thnx to
Ben)

20080607 - Ondrej Jaura

- the EXIF library exifr optionally required
- displaying EXIF basic data: exposure time, focal length and f-number
- displaying EXIF extended data: camera model
- options to enable displaying basic and extended EXIF data

- album statistics in the menu: number of photos / number of subalbums
- album statistics in the list of subalbums: total number of photos in the
album (including all subalbums)
- custom labels for no, one or more photos or albums

- automatic album highlight creation (if none is already defined) from the
first image of the album

- generator output fixed - path without delimiters / (bug #19836, thnx to Mooffie)
- typos fixed

20080423 - Mooffie

- relative links home, up and down
- explicit index.html option

* 0.5  Mind the gap

20080403 - Ondrej Jaura

- generator meta tag added (with the version number)
- support spaces in directory and file names

* 0.4  Slim and nice

20071209 - Ondrej Jaura

New settings (padding and fading) for the TripTracker viewer added.

20071130 - Michael Adams

- Ran CSS files through Flopcakes CSS optimizer: 30% reduction in sizes.
- Put in a "hack" to use only even-numbered k's on the thumbnail HTML layout: images were appearing double.

20071123 - Ondrej Jaura

Improve thumbnail quality (switches -thumbnail and -blur).

* 0.3  It is so good

20071112 - Ondrej Jaura

Author line is shown always.
Author label added to allow values such as Author, Authors, Artist, etc.
Crash if the options.yml file is missing fixed.
Typos fixed.
README style fixed and improved.
Rakefile VERSION problem fixed.

20071106 - Viktor Zigo

- other image format extensions; case insensitive (in creating thumbnails and also title generation)
- Images and albums descriptions:
  . one central file per album (description.txt containing one-line 
	mapping: file name, text separated by colon, semicolon, comma, tab.)
  . one file per image/album (the text file name is the original file name of image or directory + '.txt')
  . central description file generator (rhotoalbum.rb text)
- default album image - if no highlight given 
- author option
- alternative design - e.g. white
- parametrization: 
  . CSS to be used 
  . display title (photos, albums)
  . thumb dimensions
  . showing of title / description / date
  . copyright / license text
- clean generated script (.sh)
- copyright reworked 
- CSS adjustments more style-able
- style switcher (optional)
- XHTML strict; CSS valid
- file permissions (700) + shebang
- command line and help message
- external optional options file to override default configuration (options.yml)
- commands for recursive cleaning of generated pages (cleanindex) of any generated content (clean)

-----
* 0.2  Well, show also photos in my sub albums
20070712 - Ondrej Jaura

Only a bug fix for sub albums.
Fixed:
- bug #12232  Sub albums are not correctly generated.

-----
* 0.1  Show my photos
20070525 - Ondrej Jaura

The first public release. Still a lot of things to do, but the source code is almost
cleaned and documented.

Rhotoalbum provides the following basic functionalities: traversing subdirectories, creating
thumbnails and generating an index.html page in each subdirectory.


fx.rb
# Rhotoalbum -- a Ruby photo album generator.
#
# Copyright (C) 2007-2008  Ondrej Jaura, Viktor Zigo
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Ondrej Jaura <ondrej@valibuk.net>
# Viktor Zigo <viz@alephzarro.com>

require 'optional_require'
optional_require('rubygems')
RMAGICK_LIB = optional_require('RMagick')

def shadow anImage, opts = {}
    aBackground = opts[:background]
    anImage.background_color = '#000'
    shadow = anImage.shadow 5,5, 5.0, 0.5
    shadow.composite!(anImage, Magick::NorthWestGravity, Magick::OverCompositeOp)

    if aBackground
        blackfill = Magick::GradientFill.new(0, 0, anImage.columns, 0, aBackground, aBackground)
        background = Magick::Image.new shadow.columns, shadow.rows, blackfill
        background.composite!(shadow, Magick::NorthWestGravity, Magick::OverCompositeOp)
    else 
        shadow
    end
end

def glow anImage, opts = {}
    aBackground = opts[:background] || "none"
    cols, rows = anImage.columns, anImage.rows
    blackfill = Magick::GradientFill.new(0, 0, cols, 0, aBackground, aBackground)
    background = Magick::Image.new cols, rows, blackfill

    anImage.resize!( cols*0.89, rows*0.89 )
    shadow = anImage.clone

#    shadow.resize!( cols*0.9, rows*0.9 )
    shadow = background.composite!(shadow, Magick::CenterGravity, Magick::OverCompositeOp)
    shadow = shadow.blur_image(10, 10) 
    shadow.composite!(anImage, Magick::CenterGravity, Magick::OverCompositeOp)
end


def reflection anImage, opts = {}
    aBackground = opts[:background]
    anImage.border!(5, 5, "none")

    reflection = anImage.wet_floor 0.7, 1.0
    #reflection = reflection.distort Magick::PerspectiveDistortion, [0,0,0,0,  20,90,0,90,  90,0,90,0,  90,90,90,90], true
    reflection = reflection.blur_image(0, 3)

    parts = Magick::ImageList.new
    parts << anImage
    parts << reflection
    new = parts.append(true)
    if aBackground
        blackfill = Magick::GradientFill.new(0, 0, anImage.columns, 0, aBackground, aBackground)
        background = Magick::Image.new new.columns, new.rows, blackfill
        new = background.composite!(new, Magick::NorthWestGravity, Magick::OverCompositeOp)
    end
    new
end

def stack anImage, opts = {}
    background = opts[:background]
    images = Magick::ImageList.new
    final = Magick::Image.new( anImage.columns*2, anImage.rows*2) {
            self.background_color = background if background
        }
    5.times {
        images << rotate(anImage.clone, {:rotate_more => true}, true, true)
    }
    
    #flatten
    images.each {|i|
        final.composite!(i, Magick::CenterGravity, Magick::OverCompositeOp)
    }

    final.trim!
end

def polaroid_stack anImage, opts = {}
    background = opts[:background]
    images = Magick::ImageList.new
    final = Magick::Image.new( anImage.columns*2, anImage.rows*2) {
            self.background_color = background if background
        }
    5.times {
        images << polaroid(anImage.clone, {:rotate_more => true}, true, true)
    }
    
    #flatten
    images.each {|i|
        final.composite!(i, Magick::CenterGravity, Magick::OverCompositeOp)
    }

    final.trim!
end

def rotate anImage, opts = {}, randomize = true, shadow = false
    image = anImage

    # Bend the image
    background = opts[:background]
    image.background_color = background || "none"
    ampStrength = randomize ? (0.005*(1+rand(5)))-0.0 : 0.01
    amplitude = image.columns * ampStrength
    wavelength = image.rows  * 2

    image.rotate!(90)
    image = image.wave(amplitude, wavelength)
    image.rotate!(-90)

    tilt = randomize ? ( opts[:rotate_more] ? rand(20)-10 : rand(8)-4 ) : -4
    image.rotate!(tilt)

    if shadow
        image = shadow image,  opts
    end
    #puts "tilt:#{tilt}, yadj:#{yadj}, amp:#{ampStrength}"    
    image.background_color = background || "none"
    image.trim!
end

def polaroid anImage, opts = {}, randomize = true, shadow = true
    image = anImage
    image.border!(14, 14, "#f0f0ff")

    # Bend the image
    background = opts[:background]
    image.background_color = "none"
    ampStrength = randomize ? (0.005*(1+rand(5)))-0.0 : 0.01
    amplitude = image.columns * ampStrength
    wavelength = image.rows  * 2

    image.rotate!(90)
    image = image.wave(amplitude, wavelength)
    image.rotate!(-90)

    tilt = randomize ? ( opts[:rotate_more] ? rand(30)-15 : rand(8)-4 ) : -4
    image.rotate!(tilt)

    if shadow        
        image = shadow image,  opts
    else        
        # Make the shadow
        shadow = image.flop
        shadow = shadow.colorize(1, 1, 1, "gray75")     # shadow color can vary to taste
        shadow.background_color = "white"       # was "none"
        shadow.border!(10, 10, "white")
        shadow = shadow.blur_image(0, 3)        # shadow blurriness can vary according to taste

        # Composite image over shadow. The y-axis adjustment can vary according to taste.
        yadj = randomize ? (-amplitude/2) : (-amplitude/2)
        image = shadow.composite(image, yadj, 5, Magick::OverCompositeOp)
    end
    #puts "tilt:#{tilt}, yadj:#{yadj}, amp:#{ampStrength}"    
    image.background_color = background || "none"
    image.trim!
end

def polaroid2 anImage, opts = {}
    anImage[:caption] = "Hi!"
    anImage = anImage.polaroid { self.gravity = Magick::CenterGravity }
end

def recoverSize! anImage, anOriginalImage
    cols, rows = anOriginalImage.columns, anOriginalImage.rows
    anImage.change_geometry!("#{cols}x#{rows}") do |ncols, nrows, img|
        img.resize!(ncols, nrows)
    end
    anImage
end

def withImage aFname, anOutFname = nil
    image = Magick::Image.read(aFname).first
    new = yield image
    out = anOutFname ? anOutFname : aFname.sub(/\./, "-done.")
    puts "Writing #{out}"
    new.write(out)
end


CMDS=%w{shadow reflection polaroid stack rotate glow polaroid_stack}

def processImage! aFname, aCmd = 'shadow', opts = {}
    raise "No such command '#{aCmd}'"  unless CMDS.include?(aCmd)
    withImage(aFname, aFname ) do |img|
        originalImg = img.clone
        outImg = send aCmd, img, opts
        recoverSize! outImg, originalImg
    end
    aFname
end

if $0==__FILE__
    puts "Missing RMagick library" or exit unless RMAGICK_LIB
    fname = ARGV.shift
    cmd = ARGV.shift || 'shadow'
    color = ARGV.shift
    puts "Applying '#{cmd}'	 on the image '#{fname}'"
    puts "Usage: #{$0} <path-to-image> [#{CMDS.join('|')}] [css-background-color (e.g.\"#000000\" | red)]" or exit unless fname and CMDS.include?(cmd)

    opts = {:background=>color}

    withImage(fname, fname.sub(/\./, "-#{cmd}.") ) do |img|
        originalImg = img.clone
        outImg = send cmd, img, opts
        recoverSize! outImg, originalImg
    end 
end 
